Add a feature to IngressProcessor to resolve named ports

antonu17 opened the pull request:
## Problem statement

The Ingress processor ignores the named service port and replaces it with `80`. 

## Related Issues

closes #4322

## Testing

Added a test case to `test_fetcher.py`

## Checklist

- [x] **Does my change need to be backported to a previous release?**
  - would be nice to backport this to v2.4 as well (#4810)

- [ ] **I made sure to update `CHANGELOG.md`.**
- [x] **This is unlikely to impact how Ambassador performs at scale.**
- [x] **My change is adequately tested.**
- [ ] **I updated `DEVELOPING.md` with any any special dev tricks I had to use to work on this code efficiently.**
- [ ] **The changes in this PR have been reviewed for security concerns and adherence to security best practices.**

haq204 commented on python/ambassador/fetch/ingress.py:
```diff
@@ -221,6 +235,11 @@ def _process(self, obj: KubernetesObject) -> None:
+                try:
+                    service_port = int(service_port)
+                except:
+                    service_port = self._resolve_service_port_number(obj.namespace, service_name, service_port)
```
if `_resolve_service_port_number` fails to resolve to a port number then it would return the port name if I'm reading the above method correctly right?

So later on it would default to port 80?

antonu17 replied:
that's right, in case the port can't be resolved `_resolve_service_port_number` returns the same `service_port` value. This is possible in a situation if `Service` name specified in the `Ingress` spec doesn't exist, or if the `Service` doesn't have a port with the name provided in the `Ingress`. Both cases are not liable for any ingress controller.

haq204 reviewed (commented):
thanks for the addressing this!

Just have one question.

haq204 commented:
@antonu17 I've opened https://github.com/emissary-ingress/emissary/pull/4813 so that CI can run. Our CI is a bit funky in that only maintainers have the proper permissions to run it. 

It appears the only issue is linting. The easiest way to fix that would be to run `make format`

antonu17 commented:
> @antonu17 I've opened #4813 so that CI can run. Our CI is a bit funky in that only maintainers have the proper permissions to run it.
> 
> It appears the only issue is linting. The easiest way to fix that would be to run `make format`

Got it, pushed `make format` changes

haq204 commented:
> > @antonu17 I've opened #4813 so that CI can run. Our CI is a bit funky in that only maintainers have the proper permissions to run it.
> > It appears the only issue is linting. The easiest way to fix that would be to run `make format`
> 
> Got it, pushed `make format` changes

Looks like mypy is still not happy...
```
python/tests/unit/test_fetch.py:331: error: Argument 1 to "len" has incompatible type "Optional[Dict[str, ACResource]]"; expected "Sized"
python/tests/unit/test_fetch.py:333: error: Item "None" of "Optional[Dict[str, ACResource]]" has no attribute "values"
```

antonu17 commented:
@haq204 Thanks for re-trying CI, i see it still fails with:

```
python/tests/unit/test_fetch.py:331: error: Argument 1 to "len" has incompatible type "Optional[Dict[str, ACResource]]"; expected "Sized"
python/tests/unit/test_fetch.py:333: error: Item "None" of "Optional[Dict[str, ACResource]]" has no attribute "values"
Found 2 errors in 1 file (checked 189 source files)
```

I think I understand why. Pushed another commit that fixes it.

haq204 commented:
> @haq204 Thanks for re-trying CI, i see it still fails with:
> 
> ```
> python/tests/unit/test_fetch.py:331: error: Argument 1 to "len" has incompatible type "Optional[Dict[str, ACResource]]"; expected "Sized"
> python/tests/unit/test_fetch.py:333: error: Item "None" of "Optional[Dict[str, ACResource]]" has no attribute "values"
> Found 2 errors in 1 file (checked 189 source files)
> ```
> 
> I think I understand why. Pushed another commit that fixes it.

@antonu17 thanks! I updated the other PR for changelog updates and rebased it onto master. CI is green there so once it gets approvals from the other maintainers it should be good to merge.

> @haq204 Thanks for re-trying CI, i see it still fails with:
> 
> ```
> python/tests/unit/test_fetch.py:331: error: Argument 1 to "len" has incompatible type "Optional[Dict[str, ACResource]]"; expected "Sized"
> python/tests/unit/test_fetch.py:333: error: Item "None" of "Optional[Dict[str, ACResource]]" has no attribute "values"
> Found 2 errors in 1 file (checked 189 source files)
> ```
> 
> I think I understand why. Pushed another commit that fixes it.

antonu17 commented:
> @antonu17 thanks! I updated the other PR for changelog updates and rebased it onto master. CI is green there so once it gets approvals from the other maintainers it should be good to merge.

@haq204 thanks for your review. should I close this PR?

LanceEa commented on python/ambassador/fetch/ingress.py:
```diff
@@ -104,6 +104,20 @@ def _update_status(self, obj: KubernetesObject) -> None:
             )
 
+    def _resolve_service_port_number(self, namespace, service_name, service_port):
+        for k8s_svc in self.service_dep.discovered_services.values():
```
discovered_services is a dict, can't we just use `self.service_dep.discovered_services.get(f"{service_name}.{namespace}", None)`? 

If the cluster has a larger number of services we potentially would have to do a lot of looping here.

antonu17 replied:
My bad 🤦  will update soon

LanceEa reviewed (changes requested):
@antonu17 @haq204 - Thanks for this and this will be super useful for folks using Ingress resources. 

I left a few notes on the code and test. I need familiarize myself a little more with this part of the code base so some of my comments/questions might be obvious once I dive deeper into it.

antonu17 commented on python/ambassador/fetch/dependency.py:
```diff
@@ -35,9 +36,11 @@ class ServiceDependency(Dependency):
     """
 
     ambassador_service: Optional[KubernetesObject]
+    discovered_services: Dict[KubernetesObjectKey, KubernetesObject]
```
I saw this code for the first time in my live this week xD, sorry in advance if I misunderstood something.

So `ambassador_service` seems to be ambassador's loadbalancer service which is used for referencing in Ingress status info.
And `discovered_services` dictionary used to be a field of the `InternalServiceProcessor`, but I moved it to `ServiceDependency` class to make it available for IngressProcessor for port lookups.

antonu17 commented on python/tests/unit/test_fetch.py:
```diff
@@ -278,6 +279,64 @@ def test_mapping_v1(self):
         assert mapping.prefix == valid_mapping_v1.spec["prefix"]
         assert mapping.service == valid_mapping_v1.spec["service"]
 
+    def test_ingress_with_named_port(self):
```
Alright, I'll add more test cases to address these additional cases

antonu17 commented:
@LanceEa, @haq204 I addressed all PR comments and updated the code and tests. have a look once more please.

haq204 commented on python/ambassador/fetch/ingress.py:
```diff
@@ -105,17 +105,20 @@ def _update_status(self, obj: KubernetesObject) -> None:
+            if service_port == port.get("name", None):
+                return port.get("port", service_port)
 
+        self.logger.debug(f"Could not find port '{service_port}' in service '{service_name}'")
```
since line 119 also could also return the service port, this debug log wouldn't always get outputted if `service_port` is returned

antonu17 replied:
Correct. If the port is found we don't need to display "not found" message

haq204 reviewed (commented):
one nit but otherwise lgtm once @LanceEa takes a look

LanceEa commented on python/tests/unit/test_fetch.py:
```diff
@@ -309,33 +309,74 @@ def test_ingress_with_named_port(self):
+        mapping_missed_name = mappings.get("quote-0-3")
+        assert mapping_missed_name
+        assert mapping_missed_name.prefix == "/missed-name"
+        assert mapping_missed_name.service == "missed.default:missed"
```
maybe I'm missing something but aren't we trying to resolve the names to port numbers here? 

I would expect this one to be rejected and dropped or fallback to something, same with the metrics above too.

antonu17 replied:
Hello, @LanceEa! 
The fallback behavior is to retain the named port in the `aconf` mapping structure (this is how the ingress fetcher works at the moment). The bad port will be replaced to 80/443 later here: https://github.com/emissary-ingress/emissary/blob/aee9d485917365a2efa53ce90f121483da5c95a0/python/ambassador/ir/ircluster.py#L212-L215

antonu17 replied:
The point of these additional ingress fetcher tests is to demonstrate that port number resolution works whenever it is possible: referenced service and ports exist. In case of ingress' paths reference non-existing services or ports fetcher behaves as before.
Unfortunately, there were no tests for ingress fetcher before, so there's nothing to compare these tests with.
One thing I can do to better emphasize that the port number resolution is not guaranteed is to add the word `try` to the name of the method.

LanceEa replied:
ahh got it. Like I mentioned, I wasn't as familiar with this part of the codebase 😄. My initial gut instinct was to assume that the resolution would have happened at this stage rather than delaying until later stages. So, LGTM

haq204 commented:
@antonu17 can you rebase onto `main`? Once that is done we can run CI and get it merged

haq204 commented:
@antonu17 CI in https://github.com/emissary-ingress/emissary/pull/4813 is green and has been approved and merged.

I'm going to close this one out. 

Thanks again for your contribution!
